If origin-Header not contains port (e.g. default port), request is always a cross domain request.

jwundrak opened the pull request:
Currently Origin-Header with `http://example.com` is resolved to `getPort() = null`. 
`Uri::fromEnvironment` set the Port from $_SERVER, which should always defined, so the check, `origin->getPort() != destination->getPort()`  (in Request-Constructor) is always true.

This Pull-Request set the default Port (80 or 443), if port is not given in `HTTP_ORIGIN`

mbrodala commented:
Thanks for taking your time looking into this. The current state is intentional however as you can see in https://github.com/pagemachine/cors/commit/c6a77a4bed93410c68a5f5217d76bf41a253b76d.

The point is that if you e.g. allow access from `https://example.com` we must not modify the origin to `https://example.com:443` since the origin would not match the list of allowed origins anymore.

I'll have a look at this.

jwundrak commented:
Thanks for the fast response.
Sorry, I didn't see this commit, but I see the problem. So I add the port. Maybe this should add to the [README.md](https://github.com/pagemachine/cors/blob/master/README.md)

If you want, you can close this request.

jwundrak commented:
Sorry for comment again.

My problem is, that on a standard port, the browser does not send the port (e.g ajax-Request). I understand, that modify the ORIGIN-Url is not a good idea, so I simply extend the `isCrossOrigin` Check.

What do you think about?

mbrodala commented:
I've looked around but I couldn't find a clear statement in the specification on whether `https://example.org` and `https://example.org:443` (same for HTTP) should be seen as the same by implicit port. If we go this route the `Negotiator` must be changed to encapsulate the configured `allowedOrigins` with `Uri` objects to ensure the same normalization is applied to these too. This way the mentioned problem should not be an issue anymore.

Do you have some time to tackle this?

mbrodala commented on Classes/AccessControl/Request.php:
```diff
@@ -129,8 +129,23 @@ public function __construct(array $environment) {
+        $this->origin->getHostname() != $this->destination->getHostname();
+
+      // compare ports.
+      if ($this->isCrossOrigin === FALSE && $this->origin->getPort() != $this->destination->getPort()) {
```
Can you move this before the check above so that we have a normalized destination and origin port which could be used in the check?

mbrodala commented on Classes/AccessControl/Request.php:
```diff
@@ -129,8 +129,23 @@ public function __construct(array $environment) {
+      // compare ports.
+      if ($this->isCrossOrigin === FALSE && $this->origin->getPort() != $this->destination->getPort()) {
+        // If empty, check for standard ports
+        $destinationPort = $this->destination->getPort();
```
This block can be streamlined to a single line IMO.

mbrodala commented on Classes/AccessControl/Request.php:
```diff
@@ -129,8 +129,23 @@ public function __construct(array $environment) {
+          $originPort = $this->origin->getScheme() === 'https' ? Uri::PORT_HTTPS : Uri::PORT_HTTP;
+        }
+
+        $this->isCrossOrigin = $originPort !== $destinationPort;
```
This can be dropped then.

mbrodala commented:
@jwundrak Good idea actually, this way we avoid breaking changes. Can you make the suggested changes?

jwundrak commented:
Yes, I look for it. 
But I think the main problem is, that we get the environment information from the `$_SERVER` Array. Of course there is a port defined, but on a standard port, it is not the Uri.

Maybe we could use `GeneralUtility::getIndpEnv('TYPO3_PORT')`

mbrodala commented:
That wouldn't help, `TYPO3_PORT` is also empty if no port was set.

jwundrak commented:
Correct. So the same way like the "browser experience".

But ok, if do your suggested changes

jwundrak committed:
[BUGFIX] Use normalized ports for cross origin check

This normalizes the ports for HTTPS (443) and HTTP (80) for the
cross origin check if they where not set explicitly in origin
or destination.

jwundrak commented:
Changes applied.

mbrodala commented:
I went ahead and moved the port normalization around which slims down the request class again. This looks good from my POV now so please have another look if it fixes your problem.

jwundrak commented:
I can confirm, that your commit works here. Looks good.

mbrodala commented:
Alright, thanks for bringing this up. I'll make a release ASAP. :-)

mbrodala commented:
Version 2.0.5 is on the way.
